### PR TITLE
Add FP8 quantization for hidden state data generation

### PR DESCRIPTION
**Pull request by shubhra**
Add optional FP8 quantization of hidden states during data generation, reducing disk usage by ~50% with minimal precision loss. Training transparently dequantizes FP8 data back to bf16.

New files:
- `fp8_utils.py`: core quant/dequant with per-tensor and per-token scaling

Modified files:
- `data_generation_offline.py`: `--fp8-quantize` and `--fp8-granularity` flags
- `vllm_hidden_states_generator.py`: optional FP8 path in `generate()`
- `train/data.py`: auto-detect FP8 format and dequantize at load time

Design
- FP8 only activates when `--fp8-quantize` is passed during datagen, or when saved data contains the `fp8_format` key
- Existing bf16 data loads identically through `standardize_data_auto`

Tested with Qwen3-8B (`inference-optimization/Speculators-DEMO-Qwen3_8b`), `vLLM 0.18.0`, `temperature=0`. 

**FP8 hidden states:**
| Benchmark | k=1 | k=2 | k=3 | k=4 | k=5 |
|---|---|---|---|---|---|
| HumanEval | 1.80 | 2.43 | 2.91 | 3.17 | 3.39 |
| math_reasoning | 1.82 | 2.45 | 2.90 | 3.26 | 3.46 |
| qa | 1.74 | 2.25 | 2.67 | 2.94 | 3.09 |
| question | 1.75 | 2.27 | 2.60 | 2.85 | 2.96 |
| rag | 1.73 | 2.19 | 2.53 | 2.76 | 2.85 |
| summarization | 1.62 | 1.96 | 2.16 | 2.23 | 2.29 |
| translation | 1.65 | 2.03 | 2.26 | 2.38 | 2.44 |

**BF16 hidden states (baseline):**
| Benchmark | k=1 | k=2 | k=3 | k=4 | k=5 |
|---|---|---|---|---|---|
| HumanEval | 1.81 | 2.44 | 2.87 | 3.21 | 3.44 |
| math_reasoning | 1.82 | 2.44 | 2.91 | 3.23 | 3.51 |
| qa | 1.74 | 2.25 | 2.61 | 2.90 | 3.03 |
| question | 1.77 | 2.26 | 2.62 | 2.84 | 2.92 |
| rag | 1.73 | 2.22 | 2.57 | 2.77 | 2.82 |
| summarization | 1.62 | 1.96 | 2.14 | 2.24 | 2.31 |
| translation | 1.65 | 2.04 | 2.25 | 2.37 | 2.46 |

### CODE REVIEW

File: scripts/data_generation_offline.py [hunk 1]
```diff
@@ -118,357 +153,246 @@ def parse_args():
...
-            "By default, failed samples are skipped."
+            "Quantize hidden states to float8_e4m3fn with scaling "
+            "before saving."
         ),
     )
     parser.add_argument(
-        "--max-consecutive-errors",
-        type=int,
-        default=None,
+        "--fp8-granularity",
```
Comment (dsikka) [inline scripts/data_generation_offline.py:200]: should be called strategy, not granularity

File: src/speculators/data_generation/fp8_utils.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+import torch
+
+FP8_DTYPE = torch.float8_e4m3fn
+FP8_MAX = torch.finfo(FP8_DTYPE).max  # 448.0
+FP8_FORMAT_KEY = "fp8_format"
+FP8_FORMAT_PER_TENSOR = "per_tensor_scaled"
+FP8_FORMAT_PER_TOKEN = "per_token_scaled"
+SCALES_KEY = "hidden_states_scales"
+
+Granularity = Literal["per_tensor", "per_token"]
```
Comment (dsikka) [inline src/speculators/data_generation/fp8_utils.py:27]: It would be nice if we could be somewhat consistent with how these are defined in CT/vLLM

**Review by dsikka — COMMENTED** [review]
you'll need to rebase - the old system has been deprecated

**Comment by fynnsu** [comment]
Closing in favor of #491
